includes/class-wc-google-gtag-js.php

```diff
@@ -86,27 +86,86 @@ public function variant_data( $data, $product, $variation ) {
 	}
 
 	/**
-	 * Enqueues JavaScript to build the addImpression event
+	 * Returns Javascript string for Google Analytics events
+	 *
+	 * @param string $event The type of event
+	 * @param array  $data  Event data to be sent
+	 * @return string
+	 */
+	public static function get_event_code( string $event, array $data ): string {
+		return sprintf( "%s('event', '%s', %s)", self::tracker_var(), esc_js( $event ), self::format_event_data( $data ) );
+	}
+
+	/**
+	 * Escape and encode event data
+	 *
+	 * @param array $data Event data to processed and formatted
+	 * @return string
+	 */
+	public static function format_event_data( array $data ): string {
+		$data = apply_filters( 'woocommerce_gtag_event_data', $data );
+
+		// Recursively walk through $data array and escape all values that will be used in JS.
+		array_walk_recursive(
+			$data,
+			function( &$value, $key ) {
+				$value = esc_js( $value );
+			}
+		);
+
+		return wp_json_encode( $data );
+	}
+
+	/**
+	 * Returns an array of category names the product is atttributed to
+	 *
+	 * @param  WC_Product $_product  Product to pull info for
+	 * @return array
+	 */
+	public static function product_get_category_line( $_product ) {
+		$out            = [];
+		$variation_data = $_product->is_type( 'variation' ) ? wc_get_product_variation_attributes( $_product->get_id() ) : false;
+		$categories     = get_the_terms( $_product->get_id(), 'product_cat' );
+
+		if ( is_array( $variation_data ) && ! empty( $variation_data ) ) {
+			$parent_product = wc_get_product( $_product->get_parent_id() );
+			$categories     = get_the_terms( $parent_product->get_id(), 'product_cat' );
```

Comment:
Loading the parent product can be useful to ensure it's actually a valid product, but in this case we only use it to get the ID, which we already have. For example if `wc_get_product` would return false you'd get a fatal error here. Maybe the following would provide better fallbacks?

```suggestion
			$categories = get_the_terms( $_product->get_parent_id(), 'product_cat' );
```